findtrieentries.go

```diff
@@ -0,0 +1,299 @@
+// and returns the trie indexes (leaf indexes) of all the matches as well as the number of trie entries considered
+func findTrieKeys(log logger.Logger, massifReader MassifReader, logTenant string, trieKeys ...[]byte) ([]uint64, uint64, error) {
+
+	// get the head massif
+	headMassifContext, err := massifReader.GetHeadMassif(context.Background(), logTenant)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// find the number of massifs
+	massifCount := headMassifContext.Start.MassifIndex + 1
+
+	trieIndex := uint64(0)
+
+	matchingTrieIndexes := []uint64{}
+
+	for massifIndex := range massifCount {
+
+		massifContext, err := massifReader.GetMassif(context.Background(), logTenant, uint64(massifIndex))
+		if err != nil {
+			return nil, 0, err
+		}
+
+		// get the trieEntry count based on the size
+		// NOTE: the leaf count and trie entry count are the same
+		// NOTE: the leaf index and trie index are equivilent.
+		trieEntries := massifContext.MassifLeafCount()
+
+		log.Debugf("checking %v trie entries in massif %v for matches", trieEntries, massifIndex)
+
+		// check each leaf for matching trieKeys
+		for range trieEntries {
+
+			mmrIndex := mmr.MMRIndex(trieIndex)
+
+			logTrieKey, err := massifContext.GetTrieKey(mmrIndex)
+			if err != nil {
+				return nil, 0, err
+			}
+
+			for _, trieKey := range trieKeys {
+
+				// if a triekey matches add it to the matching trie indexes
+				if bytes.Equal(trieKey, logTrieKey) {
+					matchingTrieIndexes = append(matchingTrieIndexes, trieIndex)
+					break // only one trieKey will ever match, so if we found the matching trie key, don't keep looking
+				}
+
+			}
+
+			trieIndex++
+
+		}
+	}
+
+	// the leaf index is now the leaf size as we do an extra +1 at the end of the for loop
+	return matchingTrieIndexes, trieIndex, nil
+
+}
+
+// NewFindTrieEntriesCmd finds the trie entries associated with a given trie key in the tenants Merkle Log.
+//
+//nolint:gocognit
+func NewFindTrieEntriesCmd() *cli.Command {
+	return &cli.Command{
+		Name: "find-trie-entries",
+		Usage: `finds the matching trie entries for the given trie key.
+
+		By default returns all trie Indexes (leaf indexes) of matching trie entries.
+
+		The trieKey is HASH(DOMAIN | LOGID | APPID)
+`,
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:  logTenantFlagName,
+				Usage: fmt.Sprintf("the tenant of the log to search in. Required or can be derived from %v.", logIDFlagName),
+			},
+			&cli.StringFlag{
+				Name:  logIDFlagName,
+				Usage: fmt.Sprintf("the hexadecimal representation of the log ID. Required or can be derived from %v.", logTenantFlagName),
+			},
+			&cli.StringFlag{
+				Name:     appIDFlagName,
+				Usage:    "the app ID. For eventsv1 or assetsv2 this is the event id.",
+				Required: true,
+			},
+			&cli.Uint64Flag{
+				Name:  domainFlagName,
+				Usage: "the domain used to derive the trie key.",
+				Value: uint64(massifs.KeyTypeApplicationContent),
+			},
+			&cli.BoolFlag{
+				Name:  asMMRIndexesFlagName,
```

Comment:
I think I would be inclined to always talk in terms of mmrIndices or leafIndices. And then acknowlege trieIndex as being a proxy for leaf index. When I read trieIndex the first time, I assumed it was relative to the massif rather than the whole mountain range

Reply:
Thats fair :) will change to leafIndex

Reply:
done!